### ISSUE TITLE
Difference in type narrowing between regular and SQLAlchemy dataclasses introduced in v1.1.387

### ISSUE THREAD

**Issue by AlexanderPodorov** [microsoft/pyright#9372]
**Describe the bug**
See example code. It does not produce a warning in v1.1.386, but it does produce the warning in v1.1.387

```python
from collections.abc import Sequence
from dataclasses import dataclass

from sqlalchemy.orm import Mapped, MappedAsDataclass


@dataclass
class Delivery:
    pass


@dataclass
class Package:
    pass


@dataclass
class Shipment:
    deliveries: Sequence[Delivery]


class SQLAOrder(MappedAsDataclass):
    shipment: Mapped[Shipment | None]
    packages: Mapped[Sequence[Package]]


@dataclass
class RegularOrder:
    shipment: Shipment | None
    packages: Sequence[Package]


sqla_order = SQLAOrder(shipment=None, packages=[])

sqla_order.shipment = Shipment(deliveries=[])

# "deliveries" is not a known attribute of "None"
for _, _ in zip(sqla_order.packages, sqla_order.shipment.deliveries, strict=False):
    pass

reg_order = RegularOrder(shipment=None, packages=[])

reg_order.shipment = Shipment(deliveries=[])

# no warnings
for _, _ in zip(reg_order.packages, reg_order.shipment.deliveries, strict=False):
    pass

```

Install `SQLAlchemy==2.0.36`

**VS Code extension or command-line**
Pylance v2024.10.104, I believe the latest Pyright 1.1.387 is bundled here.
Also checked it with vanilla Pyright. No warnings observed in v1.1.386 unlike to v1.1.387.

**Comment by erictraut** [comment]
This change in behavior is caused by a [bug fix](https://github.com/microsoft/pyright/issues/9332) that was introduced in pyright 1.1.387. The new behavior is correct.

The problem is caused by the fact that you're using the `Mapped` class, which is an "asymmetric descriptor". That means its `__get__` method return type differs from the type accepted by its `__set__` method. That's an indication that its `__set__` method performs some transform on the input type, which means that it's unsafe for a type checker to assume that the same object will be returned by the `__get__` method. 

Here's the relevant type definition from the sqlalchemy code. Note that the `__set__` accepts a value of type `SQLCoreOperations[_T_co] | _T_co` but returns a value of type `_T_co`.

```python
class Mapped:
    ...
    def __get__(self, instance: object, owner: Any) -> _T_co: ...
    def __set__(self, instance: Any, value: SQLCoreOperations[_T_co] | _T_co) -> None: ...
```

When pyright detects an asymmetric descriptor, it does not apply type narrowing on assignment, as [documented here](https://microsoft.github.io/pyright/#/mypy-comparison?id=type-narrowing-for-asymmetric-descriptors).

I fixed a bug in 1.1.387 where pyright was failing to correctly detect an asymmetric descriptor when it used an overload for its `__get__` method. Now that this is fixed, pyright correctly identifies `Mapped` as an asymmetric descriptor.

You may want to discuss this with the maintainers of sqlalchemy to see if they have any thoughts on whether `Mapped` is intended to be asymmetric. If not, they could change type annotations to make it symmetric. If `Mapped` really is asymmetric, then you will need to adjust your code accordingly.
